File: ios/Runner/DeepLinkBridge.swift
import Flutter
import Foundation

/// iOS counterpart of the Android `DeepLinkBridge`. Buffers `voidtunnel://`
/// URLs emitted by the scene delegate and pushes them onto the
/// `void.deeplink/events` EventChannel. The cold-launch URL is also
/// exposed once via the `void.deeplink` MethodChannel's `consumeInitial`
/// method.
final class DeepLinkBridge: NSObject, FlutterStreamHandler {
  static let shared = DeepLinkBridge()

  private let lock = NSLock()
  private var buffer: [String] = []
  private var pendingInitial: String?
  private var sink: FlutterEventSink?

  private override init() {}

  /// Registers the method and event channels on the given binary messenger.
  /// Call once after the FlutterEngine is attached.
  func register(with messenger: FlutterBinaryMessenger) {
    let methodChannel = FlutterMethodChannel(
      name: "void.deeplink",
      binaryMessenger: messenger
    )
    methodChannel.setMethodCallHandler { [weak self] call, result in
      guard let self = self else { return }
      switch call.method {
      case "consumeInitial":
        result(self.consumeInitial())
      default:
        result(FlutterMethodNotImplemented)
      }
    }

    let eventChannel = FlutterEventChannel(
      name: "void.deeplink/events",
      binaryMessenger: messenger
    )
    eventChannel.setStreamHandler(self)
  }

  func emit(url: String, isInitial: Bool) {
    let trimmed = url.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !trimmed.isEmpty else { return }
    lock.lock()
    if isInitial {
      pendingInitial = trimmed
    }
    buffer.append(trimmed)
    if buffer.count > 8 {
      buffer.removeFirst(buffer.count - 8)
    }
    let currentSink = sink
    lock.unlock()
    if let currentSink = currentSink {
      DispatchQueue.main.async {
        currentSink(trimmed)
      }
    }
  }

  func consumeInitial() -> String? {
    lock.lock()
    defer { lock.unlock() }
    let value = pendingInitial
    pendingInitial = nil
    return value
  }

  // MARK: - FlutterStreamHandler

  func onListen(
    withArguments arguments: Any?,
    eventSink events: @escaping FlutterEventSink
  ) -> FlutterError? {
    lock.lock()
    sink = events
    let pending = buffer
    lock.unlock()
    for link in pending {
      events(link)
    }
    return nil
  }

  func onCancel(withArguments arguments: Any?) -> FlutterError? {
    lock.lock()
    sink = nil
    lock.unlock()
    return nil
  }
}

File: android/app/src/main/kotlin/com/voidtunnel/voidtunnel/DeepLinkBridge.kt
package com.voidtunnel.voidtunnel

import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.EventChannel

/**
 * Bridges incoming `voidtunnel://` deep links from the OS to the Flutter
 * side. Mirrors [VpnEventBridge]:
 *
 *  - The native bridge buffers links emitted before Flutter starts
 *    listening, so the cold-launch URL is never dropped.
 *  - The cold-launch URL is also exposed once via the companion
 *    `void.deeplink` MethodChannel's `consumeInitial` method; the Flutter
 *    side decides which side to consume on.
 *
 * Lives as a process-wide singleton because MainActivity may be recreated
 * (configuration changes, deep-link relaunch) while the same FlutterEngine
 * is in use.
 */
object DeepLinkBridge : EventChannel.StreamHandler {
    private val mainHandler = Handler(Looper.getMainLooper())
    private const val MAX_BUFFER = 8

    @Volatile
    private var sink: EventChannel.EventSink? = null
    private val buffer: ArrayDeque<String> = ArrayDeque()
    private var pendingInitial: String? = null

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        sink = events
        synchronized(buffer) {
            for (link in buffer) {
                events?.success(link)
            }
        }
    }

    override fun onCancel(arguments: Any?) {
        sink = null
    }

    /**
     * Emit a deep link from native code. If [isInitial] is true the link is
     * also stashed for the next [consumeInitial] call, so a cold-launch URL
     * can be retrieved either through the event stream or the one-shot
     * method, depending on which Flutter listener wins the race.
     */
    fun emit(link: String, isInitial: Boolean = false) {
        if (link.isBlank()) return
        synchronized(buffer) {
            buffer.addLast(link)
            while (buffer.size > MAX_BUFFER) {
                buffer.removeFirst()
            }
            if (isInitial) {
                pendingInitial = link
            }
        }
        mainHandler.post {
            sink?.success(link)
        }
    }

    /**
     * One-shot read of the cold-launch URL. Returns null on subsequent
     * calls until another initial URL is emitted.
     */
    fun consumeInitial(): String? {
        synchronized(buffer) {
            val value = pendingInitial
            pendingInitial = null
            return value
        }
    }
}
